### DISCUSSION TITLE
Why are global components not recommended?

### DISCUSSION THREAD

**Discussion by justanothersynth** [nuxt/nuxt#24191] [Questions]
Under the [Dynamic Components](https://nuxt.com/docs/guide/directory-structure/components#dynamic-components) section of the Nuxt 3 docs, it is stated:

> Alternatively, though **not recommended**, you can register all your components globally, which will create async chunks for all your components and make them available throughout your application.

Why is this not recommended?

**Comment by woldtwerk** [comment]
I would like to know this as well. It could result in an excessive amount of chunks. I don't see this being a problem in most projects though.

**Comment by guibulator** [comment]
I think that's the only reason,  but as the project grows, it might impact loading time of your app.  I personally prefer to import from `#components` when using dynamic components and not rely on global components.  If you are building enterprise applications only available inside your own network, that's probably not big of a deal.

**Comment by m0k99** [comment]
Global components in Vue.js are not inherently bad, but they can lead to several issues if not used judiciously:

- **Performance Impact**: Registering too many global components can affect the initial loading time of your application, especially if those components aren't immediately needed. Lazy-loading components can mitigate this issue.
- **Potential Conflicts**: Global components might inadvertently share names with components from third-party libraries or future additions to Vue itself. This can lead to conflicts and unexpected behavior.

Instead of using global components extensively, it's often recommended to use local components scoped to the parent component where they are needed. This keeps the components encapsulated and easier to reason about. When a component is only used within a specific parent component, there's little benefit to making it global.

However, there are cases where global components are appropriate, such as utility components or components used throughout the entire application (e.g., layout components). In such cases, it's essential to strike a balance and use global components judiciously, considering the trade-offs mentioned above.

**Reply by lateek35** [reply]
@m0k99 Thanks for this explanation. So how should we handle like for example, a page builder ? So an user can build pages by assembling component.

So in our nuxt, we should have something like :
```
<ComponentResolver :components="componentListFormApi" />
```

Because we can't predict the possible component wanted, we need to register all of them globally, which goes against the performance impact you mentioned.
So not sure how to approach this, by the same time, tools like storyblok [seems to work like that](https://github.com/storyblok/storyblok-vue?tab=readme-ov-file#short-form)

**Reply by jwhazel** [reply]
@lateek35 I'm currently experimenting with this, but also looking around to see if someone else has come up with a better solution. Sharing what I've got so far: 

My initial idea was to run a script on a pre-build hook that pulled page data/sitemap from an api, then programmatically generated the vue files in the `/pages` folder, then delete them after build. Although I feel it's super clunky and I kind of hate it, it should work since it basically mimics a non dynamic site.

I'm also looking to see if there is way with Nuxt's api that I could pass raw data about the page instead of having to actually write any vue files in the `/pages`. This gave way to another, grander idea - if this api did exist, then could we also tap into per-page build hooks themselves? Effectively intercepting the component dependency list in order to manually tree shake them. TBD.
